File: experiments/006-SmalltalkVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260915-110924-39025/workdir/src/main/java/customerimporter/Identification.java
package customerimporter;

import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

public abstract class Identification {

    private static final Map<String, Function<String, Identification>> creatorsByType =
        Map.of(DNI.TYPE, DNI::new,
               CUIT.TYPE, CUIT::new);

    private final String number;

    // instance creation

    public static Identification forTypeAndNumber(String anIdentificationType, String anIdentificationNumber) {
        Identification newIdentification = creatorForType(anIdentificationType).apply(anIdentificationNumber);
        newIdentification.assertIsValid();
        return newIdentification;
    }

    // error messages

    public static String invalidIdentificationTypeErrorDescription() {
        return "Invalid identification type";
    }

    // instance creation - private

    private static Function<String, Identification> creatorForType(String anIdentificationType) {
        return Optional.ofNullable(creatorsByType.get(anIdentificationType))
            .orElseThrow(() -> new RuntimeException(invalidIdentificationTypeErrorDescription()));
    }

    // initialization

    protected Identification(String aNumber) {
        number = aNumber;
    }

    // identification

    public abstract String type();

    public String number() {
        return number;
    }

    public boolean is(String anIdentificationType, String anIdentificationNumber) {
        return type().equals(anIdentificationType) && number().equals(anIdentificationNumber);
    }

    // type

    public boolean isDNI() {
        return false;
    }

    public boolean isCUIT() {
        return false;
    }

    public Object dniNumberIfNone(Supplier<Object> aNoneBlock) {
        return aNoneBlock.get();
    }

    public Object cuitNumberIfNone(Supplier<Object> aNoneBlock) {
        return aNoneBlock.get();
    }

    // validation

    protected abstract void assertIsValid();

    protected abstract String invalidNumberErrorDescription();

    protected void assertThat(boolean aCondition) {
        if (!aCondition) throw new RuntimeException(invalidNumberErrorDescription());
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-205812-95418/output/project/src/customerimporter/identification.py
from abc import ABC, abstractmethod


class Identification(ABC):

    _number = None

    # instance creation

    @classmethod
    def for_(cls, an_identification_type, an_identification_number):
        for an_identification_class in cls.__subclasses__():
            if an_identification_class.is_for(an_identification_type): return an_identification_class.with_number(an_identification_number)
        raise RuntimeError(cls.invalid_identification_type_error_description())

    @classmethod
    def with_number(cls, an_identification_number):
        cls.assert_is_valid_number(an_identification_number)
        return cls(an_identification_number)

    # error messages

    @classmethod
    def invalid_identification_type_error_description(cls):
        return "Invalid identification type"

    @classmethod
    @abstractmethod
    def invalid_number_error_description(cls):
        pass

    # initialization

    def __init__(self, an_identification_number):
        self._number = an_identification_number

    # type

    @classmethod
    @abstractmethod
    def identification_type(cls):
        pass

    @classmethod
    def is_for(cls, an_identification_type):
        return cls.identification_type() == an_identification_type

    # number - validation

    @classmethod
    @abstractmethod
    def assert_is_valid_number(cls, an_identification_number):
        pass

    @classmethod
    def assert_valid_number(cls, a_condition):
        if not a_condition: raise RuntimeError(cls.invalid_number_error_description())

    # number

    def number(self):
        return self._number

    @abstractmethod
    def cuit_number_if_none(self, a_none_block):
        pass

    @abstractmethod
    def dni_number_if_none(self, a_none_block):
        pass

    @abstractmethod
    def is_cuit(self):
        pass

    @abstractmethod
    def is_dni(self):
        pass
